# Analyzing ISIS tweets Dataset

A one-hour trainining of analysing ISIS tweet dataset from kaggle - https://www.kaggle.com/kzaman/how-isis-uses-twitter

This Notebook implements text analytics process:
* tokenizing
* stemming
* stop-words removal
* index and inverted index creation for token-count


In [2]:
import pandas as pd
import nltk
import string
import time
import math
from tqdm import tqdm
from sklearn.cross_validation import train_test_split

C:\Users\22334\AppData\Local\Continuum\Anaconda27\lib\site-packages\sklearn\cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [4]:
data = pd.read_excel('tweets.xlsx')
data


name         username  \
0           GunsandCoffee  GunsandCoffee70   
1           GunsandCoffee  GunsandCoffee70   
2           GunsandCoffee  GunsandCoffee70   
3           GunsandCoffee  GunsandCoffee70   
4           GunsandCoffee  GunsandCoffee70   
5           GunsandCoffee  GunsandCoffee70   
6           GunsandCoffee  GunsandCoffee70   
7           GunsandCoffee  GunsandCoffee70   
8           GunsandCoffee  GunsandCoffee70   
9           GunsandCoffee  GunsandCoffee70   
10          GunsandCoffee  GunsandCoffee70   
11          GunsandCoffee  GunsandCoffee70   
12          GunsandCoffee  GunsandCoffee70   
13          GunsandCoffee  GunsandCoffee70   
14          GunsandCoffee  GunsandCoffee70   
15          GunsandCoffee  GunsandCoffee70   
16          GunsandCoffee  GunsandCoffee70   
17          GunsandCoffee  GunsandCoffee70   
18          GunsandCoffee  GunsandCoffee70   
19          GunsandCoffee  GunsandCoffee70   
20          GunsandCoffee  GunsandCoffee70   
21          GunsandCoffee  GunsandCoffee70   
22          GunsandCoffee  GunsandCoffee70   
23          GunsandCoffee  GunsandCoffee70   
24          GunsandCoffee  GunsandCoffee70   
25          GunsandCoffee  GunsandCoffee70   
26          GunsandCoffee  GunsandCoffee70   
27          GunsandCoffee  GunsandCoffee70   
28          GunsandCoffee  GunsandCoffee70   
29     Abu Layth Al Hindi  AbuLaythAlHindi   
...                   ...              ...   
17380                  ☝️       melvynlion   
17381           N i d a l      Nidalgazaui   
17382         Anaksabil97          nvor85j   
17383           N i d a l      Nidalgazaui   
17384        Mountain Man   Mountainjjoool   
17385         Anaksabil97          nvor85j   
17386           N i d a l      Nidalgazaui   
17387        كتكات كوكونت    Uncle_SamCoco   
17388         Anaksabil97          nvor85j   
17389        كتكات كوكونت    Uncle_SamCoco   
17390         Anaksabil97          nvor85j   
17391        كتكات كوكونت    Uncle_SamCoco   
17392        كتكات كوكونت    Uncle_SamCoco   
17393    one who believes           ro34th   
17394      Musa Al-Ansari        lNSlDEWAR   
17395        كتكات كوكونت    Uncle_SamCoco   
17396        كتكات كوكونت    Uncle_SamCoco   
17397        كتكات كوكونت    Uncle_SamCoco   
17398        كتكات كوكونت    Uncle_SamCoco   
17399        كتكات كوكونت    Uncle_SamCoco   
17400        Mountain Man   Mountainjjoool   
17401        Mountain Man   Mountainjjoool   
17402        كتكات كوكونت    Uncle_SamCoco   
17403         Anaksabil97          nvor85j   
17404        كتكات كوكونت    Uncle_SamCoco   
17405         Anaksabil97          nvor85j   
17406         Anaksabil97          nvor85j   
17407         Anaksabil97          nvor85j   
17408         Anaksabil97          nvor85j   
17409    one who believes           ro34th   

                                             description           location  \
0           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
1           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
2           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
3           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
4           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
5           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
6           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
7           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
8           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
9           ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
10          ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
11          ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
12          ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews                NaN   
13          ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews      

## extract Tokens from text

In [6]:
def tokenize(msg):
    #converts a text into a list of stemmed , non-stop word, tokens
    tokenizer = nltk.tokenize.RegexpTokenizer(r'\w+') # stop character sensitive tokenizer
    tok = nltk.word_tokenize(msg) # white space tokenizes
    stemmer = nltk.stem.porter.PorterStemmer() #Porter stemmer object
    out = [] # result list
    for i in tok:
        # for tokens contains digits - we'll take them as-is, while other tokens will be splitted using puctuation
        if any(c.isdigit() for c in i): #token contains numbers
            steminput = [i] # add the token to stemmer input
        else: # non-digits
            steminput = tokenizer.tokenize(i) #split using punctuation
        for mem in steminput:
            w = stemmer.stem(mem.lower()) # w is stemmed lowercase token
            if w in nltk.corpus.stopwords.words('english') or w in list(string.punctuation): continue
            # stop words and sole-punctuation tokens removal
            else: out.append(w) #add the final token to the output
    return out # list of stemmed , non-stop word, tokens

In [7]:
data['token'] = data.apply(lambda x: tokenize(x['tweets']) ,axis = 1)

In [10]:
tw= data.iloc[:,[1,8]].groupby('username').sum().reset_index() # user and his words

In [11]:
twavgstatuses =  data.iloc[:,[1,4,5]].groupby('username').max().reset_index()

In [13]:
tw['num'] =  twavgstatuses['numberstatuses']
tw['follow'] = twavgstatuses['followers']
tw.head()

username                                              token  num  follow
0    04_8_1437  [amaqag, martyrdom, oper, rig, vehicl, hit, ga...   56      71
1  06230550_IS  [current, situat, map, wilayat, kheyr, close, ...   93     140
2    1515Ummah  [asalam, alikum, la, familia, isi, dawlah, isl...  169     214
3  1Dawlah_III  [wilayatninawa, tour, sukar, neighborhood, mos...  333     632
4     432Mryam  [pleas, follow, support, http, //t.co/g2bqagw3...   70     169

# Build index for tf-idf

In [14]:
def createIndex(data):
    #builds inverted index based on documents
    index = {"$ N $":0} # amount of documents in corpus
    docs = {} # regular index of documents and included words
    print "initializing index of",len(data),"documents" ,time.ctime(), 
    for i in data:
        index["$ N $"]+=1
        tfdic = tf_calc(readDocument(path,i))
        docs[i] = tfdic # add document to documents index
        for w in tfdic:
            if index.has_key(w):
                index[w][i] = tfdic[w] # add document to the inverted index
            else:
                index[w] = {i:tfdic[w]}
                
    print "index created with",len(index),"words", time.ctime()
    return index,docs

In [15]:
def tf_calc(wordList):
    # reads list of words and returns their tf
    total = len(wordList)
    out = {}
    for w in wordList:
        if out.has_key(w):
            out[w]+= 1.0/total # add 1 word to the counter
        else:
            out[w] = 1.0/total
    return out # dictionary of tf rates

In [16]:
def createIndex(docs):
    #builds inverted index based on documents
    index = {"$ N $":0} # amount of documents in corpus
    print "initializing index of",len(docs),"users" ,time.ctime(), 
    for i in docs:
        index["$ N $"]+=1
        tfdic = docs[i]
        for w in tfdic:
            if index.has_key(w):
                index[w][i] = tfdic[w] # add document to the inverted index
            else:
                index[w] = {i:tfdic[w]}
    print    
    print "index created with",len(index),"words", time.ctime()
    return index

# Tf - idf implementation

In [17]:
def tfidf_calc(doc, word, index):
    # tf-idf calculation based on document and a word
    if doc.has_key(word) == False or index.has_key(word)== False:
        return 0 # if there are no suitable documents return 0
    tf = doc[word]*1.0 # tf based on the document
    df = len(index[word])
    idf = math.log(index['$ N $']*1.0/df,10) # idf based on global df
    return tf*idf

In [18]:
def getVectorSize(v, index):
    # calculates vector size (based on tf-idf)
    ans = 0
    for i in v:
        ans += math.pow(tfidf_calc(v,i,index),2)
    return math.sqrt(ans)

def getSimilarity(d,q, index):
    # calculate cosine similarity between a query and a document
    ans = 0
    for i in q:
        ans += tfidf_calc(q,i,index) * tfidf_calc(d,i,index)
    
    return ans/(getVectorSize(q, index)*getVectorSize(d, index))

In [19]:
# Find top K similar tweets

In [20]:
def TopKSimilarity(q,t,index,k):
    # get k most similar documents
    similarityDict = {}
    for doc in t:
        similarityDict[doc] = getSimilarity(t[doc],q,index)
    Topk =  sorted(similarityDict, key=similarityDict.get, reverse=True)[:k]
    return Topk # return list of relevant docs

## Popularity Prediction  - KNN regression

In [21]:
train, test = train_test_split(tw, test_size = 0.3)

In [22]:
# building index on the train
users = {}
for i in tqdm(train.iterrows()):
    users[i[1][0]] = tf_calc(i[1][1])

index = createIndex(users)

78it [00:00, 928.57it/s]


initializing index of 78 users Sun Jun 04 17:25:54 2017
index created with 22769 words Sun Jun 04 17:25:54 2017


In [23]:
def predict(user,users,index,k,train):
    topK = TopKSimilarity(user,users,index,k)
    scores = train[train['username'].isin(topK)].median()
    num,follow = scores['num'], scores['follow']    
    return (num,follow)

In [24]:
testusers = {}
for i in tqdm(test.iterrows()):
    testusers[i[1][0]] = tf_calc(i[1][1])

34it [00:00, 739.13it/s]


In [25]:
ks = [1,3,5,7,9]
for k in tqdm(ks):
    test.loc[:,'pnum'+str(k)] = test.apply(lambda x: predict(testusers[x['username']],users,index,k,train)[0] ,axis = 1)
    test.loc[:,'pfollow'+str(k)] = test.apply(lambda x: predict(testusers[x['username']],users,index,k,train)[1] ,axis = 1)

  0%|                                                    | 0/5 [00:00<?, ?it/s]C:\Users\22334\AppData\Local\Continuum\Anaconda27\lib\site-packages\pandas\core\indexing.py:297: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
C:\Users\22334\AppData\Local\Continuum\Anaconda27\lib\site-packages\pandas\core\indexing.py:477: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s
100%|████████████████████████████████████████████| 5/5 [01:57<00:00, 23.44s/it]


# Results for differents K's

In [26]:
test.head()

username                                              token  \
103   squadsquaaaaad  [bismillah, acc, number, 45, follow, rt, jzk, ...   
88         lNSlDEWAR  [isi, know, ca, n, move, forward, fastli, larg...   
56      _IshfaqAhmad  [game, come, espncricinfo, ellys, perri, beaut...   
12      Abu_Ibn_Taha  [rt, wikileak, hillari, email, egyptian, milit...   
105  thefIamesofhaqq  [rt, eplc24, alert, gt, hijrahwit, plz, bewar,...   

       num  follow   pnum1  pfollow1   pnum3  pfollow3   pnum5  pfollow5  \
103     60      57   847.0     353.0   847.0     353.0   847.0     353.0   
88     259     528  8473.0     965.0  8473.0    7566.0  8473.0    2230.0   
56   13960    1667  8473.0     965.0  7374.0    2230.0  7374.0    2230.0   
12      24      73  8473.0     965.0  8473.0    1482.0  8473.0    2230.0   
105    224     318  8473.0     965.0  8473.0     965.0   211.0     754.0   

      pnum7  pfollow7   pnum9  pfollow9  
103   847.0     353.0   275.0     353.0  
88   7374.0    2021.0  7374.0    2021.0  
56   5241.0    2021.0   807.0    1307.0  
12   7374.0    2021.0  5241.0    1482.0  
105  5241.0     965.0  5241.0     965.0

# RMSE evaluation

In [28]:
from sklearn.metrics import mean_squared_error
y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0.0, 2, 8]
mean_squared_error(y_true, y_pred)

0.375

In [41]:
res = {}
for k in ks:
    res[k]=  (mean_squared_error(test['num'],test['pnum'+str(k)])**0.5, mean_squared_error(test['follow'],test['pfollow'+str(k)])**0.5)
res

{1: (10381.872764016871, 886.27477149484253),
 3: (7493.1455854471351, 3808.2648125298406),
 5: (7044.245500865979, 1347.8700626849582),
 7: (6632.943169380791, 1121.900095164187),
 9: (6211.0973975902953, 1050.3175850245318)}

<b> best k for num (number of tweets) - k=9, RMSE = 6211

best k for follow - k=1, RMSE = 886

# Single user example

In [230]:
line = test.iloc[1,:]
print line

username                                              nvor85j
token       [dowlty__isi, jazakallah, khairan, akhi, break...
num                                                       133
follow                                                    238
Name: 97, dtype: object


In [259]:
top5 = TopKSimilarity(u,users,index,5)
print(top5)
#scores = train[train['username'].isin(topK)].mean()
#num,follow = scores['num'], scores['follow'] 

[u'ro34th', u'btt_ar', u'k_kid04', u'pleaoftheummah', u'al_zaishan10']


In [250]:
train[train['username'].isin(topK)].median()

num       179
follow    187
dtype: float64